packages/lesswrong/server/repos/UsersRepo.ts

```diff
@@ -284,6 +284,32 @@ export default class UsersRepo extends AbstractRepo<DbUser> {
     `)
   }  
 
+  async getUsersWithNewDialogueChecks(hours: number): Promise<DbUser[]> {
+    return this.manyOrNone(`
+      SELECT *
+      FROM "Users"
+      WHERE EXISTS (
+        SELECT 1
+        FROM "DialogueChecks"
+        WHERE "DialogueChecks"."checkedAt"
+        > (
+            SELECT GREATEST(MAX("checkedAt"), NOW() - INTERVAL '$1 hours')
+            FROM "DialogueChecks"
+            WHERE "userId" = "Users"._id
+        )
+        AND "targetUserId" = "Users"._id
+        AND "checked" = true
+        AND NOT EXISTS (
+          SELECT 1
+          FROM "DialogueChecks"
+          WHERE "userId" = "Users"._id
+          AND "targetUserId" = "DialogueChecks"."userId"
+          AND "checked" = true
+        )
+      )
+    `, [hours])
+  }
```

Comment:
I just realized that all the `checkedAt` complexity might be unnecessary... I think you can just check whether `createdAt` was in the last [period of time] and whether `checked` is still true.

Reply:
I also suspect `WHERE EXISTS` might not perform amazingly well, but maybe the query planner is smarter than I am.  I would just do it with a join, which starts by getting all the unidirectional checks that were created in the last hour, and then joins on target users.

Reply:
I think I prefer allowing:

1. Robert checks Rafe on 1st December
2. then unchecks him 10 mins later, realising he doesn't want to dialogue with him at that point
3. then checks Rafe on 20th December, ready now for a dialogue
4. So Rafe gets a notification

Which, IIUC, the `checkedAt` approach does, but `createdAt` doesn't

Reply:
I suppose naïvely, we're quadratic in the "WHERE EXISTS". My guess is the quesry planner is smarter than that, but seems worth rewriting